### DISCUSSION TITLE
Add nix support to the project?

### DISCUSSION THREAD

**Discussion by jarmos-san** [developersIndia/developersIndia#32] [Ideas]
This discussion thread hopes to initiate a productive conversation on supporting the [`nix` package manager](https://en.wikipedia.org/wiki/Nix_package_manager) for this specific project.

@siddhantk232 please continue the discussions we had on Discord & explain the following questions:

1. Why & how does `nix` support help maintain the project?
2. For the uninitiated (like me), how're we supposed to get started with using `nix` for these project?
3. Can you share a detailed workflow we'll need to add `nix` support to the project & would you like to contribute to the project?
4. If we add `nix` support to the project, will it hamper other established workflows of other contributors & maintainers who doesn't use `nix?

Consider this thread an open-ended discussion, so please keep contributing your ideas, knowledge, resources & anything else you think might be helpful to it.

CC: @Bhupesh-V @varshney-himanshu

**Comment by siddhantk232** [comment]
While I strongly believe that nix is a useful tool, I don't think we need it at this stage of developersindia.in project. I would still love to answer the questions about nix:

1. `nix` can help in a number of ways, from development to deployment. One of its major selling points is reproducible environments. This can help us maintain a declarative build environment that is trackable through VCS. Also, `nix` is language-agnostic, so we can use it for other languages/tools for this project. For example, if someone from the community wants to build automation scripts in Python and want some dependencies from `pip`, `nix` will be able to do that (we'll be able to pin specific Python version and `pip` deps). All this with the guarantee that it's reproducible on any machine so other contributors will have the same developer experience. ​


2.  Getting started with `nix` is easy but it requires a bit of a learning curve. Please refer to the [official guides](https://nixos.org/) to learn more about [getting started](https://nixos.org/learn.html). One thing I would like to clarify tho, `nix` means three different things, the Nix expression language, the Nix package manager (aka nixpkgs), and the NixOS (which is an OS build on top of nix). Nix generally refers to the Nix package manager and is available on any system (Linux, MacOS, Windows through WSL)

3. To get started in a dev environment, we typically need a `shell.nix`, which, the nix package manager uses to create a shell with all the dependencies available. A simple `shell.nix` that uses `nodejs-16`, `git`, `neovim`, and `yarn` would look like:

    ```nix
      { mkShell, pkgs }:
      mkShell {
        name = "developerindia-dev";
        buildInputs = with pkgs; [
          nodejs-16_x
          yarn
      
          # tools
          git
          neovim # vscode is also available
        ];
      }
    ```
    These tools will simply be available to the user by running:

    ```shell
      gh repo clone developersindia/developersindia.in
      cd developersindia.in
      nix-shell
    ```

    > Would you like to contribute to the project?

    NO!

4. Nix, in development, will launch a shell session with all the tools mentioned in `shell.nix` available for use. Nothing will change for those who don't want to use nix. They'll still be able to use tools/deps installed on their system just as usual.

#### Some links to know more about nix

- The [NixOS](https://nixos.org) website
- The [official docs](https://nixos.org/manual/nix/stable/) on using `nix` available on the NixOS website.
- GitHub repository of [nixpkgs](https://github.com/nixos/nixpkgs)
- [Talk: How Shopify Uses `nix`](https://www.youtube.com/watch?v=KaIRpx11qrc)

**Reply by jarmos-san** [reply]
Thanks for sharing some resources & a brief about supporting `nix` for this project.
